internal/cmd/headless_repl/src/processes/invoke_go.rs

```diff
@@ -0,0 +1,130 @@
+use crate::invoke_go::OutputError::InvalidMethod;
+use crate::processes::lsp_invoke::add_headers;
+use crate::processes::process_completion::process_completions_response;
+use crate::LSPSuggestionHelper::LSPSuggestionHelper;
+use crate::{CommandHint, MyHelper};
+use lsp_types::{
+    DidChangeTextDocumentParams, TextDocumentContentChangeEvent, Url,
+    VersionedTextDocumentIdentifier,
+};
+use regex::Regex;
+use rustyline::Helper;
+use serde_json::Value;
+use std::collections::HashSet;
+use std::env::consts::OS;
+use std::fmt::format;
+use std::hash::Hash;
+use std::io::{stdin, Read};
+use std::process::{Child, ChildStdout, Command, Stdio};
+use std::ptr::write;
+use std::str;
+use std::str::from_utf8;
+use std::string::String;
+use std::sync::atomic::AtomicUsize;
+use std::sync::mpsc::Sender;
+use std::sync::{Arc, Mutex, RwLock};
+use tower_lsp::jsonrpc;
+use tower_lsp::jsonrpc::RequestBuilder;
+
+pub fn start_go() -> Child {
+    let mut child = Command::new("./main")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .spawn()
+        .expect("failure to execute");
+    child
+}
+
+#[derive(Debug)]
+pub enum OutputError {
+    InvalidMethod,
+}
+
+impl From<serde_json::Error> for OutputError {
+    fn from(_: serde_json::Error) -> Self {
+        InvalidMethod
+    }
+}
+
+pub fn form_output(request_type: &str, text: &str) -> Result<String, OutputError> {
+    match request_type {
+        "Service.DidOutput" => {
+            let mut cleaned = text.to_string().replace("\"", "\\\"");
+            let mut param = (r#"[{"input": "#).to_string();
+            let mut other_side = format!(r#""{}""#, cleaned);
+            other_side.push_str("}]");
+            param.push_str(other_side.as_str());
+            let paramm: Value =
+                serde_json::from_str(param.as_str()).expect("failure going to value");
+
+            let req: RequestBuilder = jsonrpc::Request::build("Service.DidOutput").params(paramm);
+            let a = serde_json::to_value(req.finish())?;
+            let res = serde_json::to_string(&a).unwrap();
+
+            // println!("{}", res);
+            Ok(res)
+        }
+        _ => Err(InvalidMethod),
+    }
+}
+
+pub fn read_json_rpc(child_stdout: ChildStdout, mut storage: Arc<RwLock<HashSet<CommandHint>>>) {
+    let re = Regex::new(r"Content-Length: ").unwrap();
```

Comment:
You don't actually need to do all this parsing, the RPC format is basically HTTP headers followed by a HTTP body. So either pull in a LSP dependency to do the parsing for you or just piggy back on a HTTP parser https://docs.rs/httparse/latest/httparse/fn.parse_headers.html .

Alternatively you can use https://doc.rust-lang.org/std/io/trait.BufRead.html#method.read_line to get a full line and then you ought to be able to more easily work with it instead of decoding one digit at a time.

Reply:
Weston looked at using `httparse` but it seemed like for there to be a real win, we probably want to make a bigger change.

My hope is we can later lift out the transport and client code from `lspower` almost verbatim. They only have an implementation for Server->Client comms, but we should be able to build our own Client->Server version based closely on theirs.

Similarly, I'd like to try and rework the transport for talking to the REPL backend to rely on a library if we can find a good "plain jsonrpc" client. https://github.com/paritytech/jsonrpsee seems to only have http and websocket transports, but we might be able to implement our own for dealing with stdin/out. Alternatively, a move to gRPC might make sense since it would afford us some interesting cloud opportunities.